portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -29,73 +30,87 @@
-      }
-
-      response.setContentType("application/json;");
-      response.getWriter().println(convertToJsonUsingGson(messages));
-  }
-
-  @Override
-  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-      String clientMessage = request.getParameter("clientMessage");
-      long timeStamp = System.currentTimeMillis();
-
-      Entity messageEntity = new Entity("Messages");
-      messageEntity.setProperty("message", clientMessage);
-      messageEntity.setProperty("timestamp", timeStamp);
-
-      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-      datastore.put(messageEntity);
+    private List<String> messages;
+    private List<Comments> messagesEntityList;
+    private int amtOfComments;
+    private static final String NAME = "name";
+    private static final String EMAIL = "email";
+    private static final String CLIENTMESSAGE = "clientMessage";
+    private static final String TIMESTAMP = "timestamp";
+
+    public void init() {
+        messages = new ArrayList<String>();
+        messagesEntityList = new ArrayList<Comments>();
+        amtOfComments = 0;
+    }
 
-      response.sendRedirect("/index.html");
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response)
+        throws IOException {
+
+        //TO-DO:figure out how to sort queries based on timestamp
+        Query query = new Query("Messages");
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        PreparedQuery results = datastore.prepare(query);
+        List<Entity> limitedResults = results.asList(FetchOptions.Builder.withLimit(amtOfComments));
+
+        for (Entity entity : limitedResults) {
+            long id = entity.getKey().getId();
+            String name = (String) entity.getProperty(NAME);
+            String email = (String) entity.getProperty(EMAIL);
+            String message = (String) entity.getProperty(CLIENTMESSAGE);
+            long timeStamp = (long) entity.getProperty(TIMESTAMP);
+
+            Comments commentsObject = new Comments(id, name, email, message, timeStamp);
+            messagesEntityList.add(commentsObject);
+        }
+
+        response.setContentType("application/json;");
+        response.getWriter().println(convertToJsonUsingGson(messagesEntityList));
     }
 
-  private String convertToJsonUsingGson(List<String> messageList) {
-      Gson gson = new Gson();
-      String json = gson.toJson(messageList);
-    
-      return json;
-  }
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException { 
+        String name = request.getParameter(NAME);
+        String email = request.getParameter(EMAIL);
+        String clientMessage = request.getParameter(CLIENTMESSAGE);
+        long timeStamp = System.currentTimeMillis();
+        amtOfComments = Integer.parseInt(request.getParameter("commentAmount"));
```

Comment:
when user post a comment, why they need to pass in "amount of comments"? logically speaking, isn't that a parameter for get request such as I am getting 10 messages.

Reply:
When a user posts a comment, there is an input box that asks them how many comments they would like to show. So commentAmount is an input value.

Reply:
I am trying to understand the intention behind this design. Correct me if I am wrong, there are two workflows:

1.  posting a comment - it means user adds a comment to your page. We just need to add this comment to our server and display it on our page.
2. retrieving comments - when user browse through all comments, we ask them how many comments they want to see at a time.

If above workflow is correct, amountOfComments is only for "doGet".

Reply:
The intention behind the design is that before you post a comment, no comments are shown. When you are submitting a comment, the client has 4 fields: name, email, message, and the # of comments you want to show AFTER you click the submit button. So the # of comments input is sent along with the name, email, and message. So all the information is sent at the same time.

Reply:
My take on posting a comment is that you just display the newly added message. So current design is kind of mix of "doPost" and "doGet"?

Let's see if Kenneth has a better idea.

Reply:
Yes you are correct, we might be thinking of different things though? I can explain tomorrow if you want

This might help clear it up though: there is only one workflow. once the user hits "submit" comments are shown with the specified number of comments that the user stated when they pressed submit. are you suggesting I use two different input forms?

Reply:
I see. let's hop on a call tomorrow. That should be easier.

Reply:
@kennzoid

Reply:
I agree with Tong that the user experience is a bit weird with the current version. I think of the comment count filter as being a part of the 'fetching comments' idea and pretty separate from the 'post a comment' idea. I also think it'd be better to show comments on page load and not after submitting a comment.

But I think what Kofi is saying is that the structure that's given by the exercise (which is one doGet method and one doPost method) doesn't really lend itself to that. I don't think we need to change the exercise dramatically, let's just do the best we can with the current structure and then scrutinize the user experience more when we get to the capstone project.

Reply:
Ha, yeah, you are right, Kenneth. If the exercise leads us to this point, I am okay to approve this. Also, as long as Kofi knows the difference between each design, I will be happy too as the main purpose of starter project is to learn.